File: packages/sdk/src/jwt-utils.ts
export interface DecodedFluxyJwt {
  exp?: number;
  sub?: string;
  tid?: string;
}

/** Decode payload from a JWT (no signature verification). */
export function decodeFluxyJwtPayload(token: string): DecodedFluxyJwt {
  try {
    const part = token.split(".")[1];
    if (!part) return {};
    const normalized = part.replace(/-/g, "+").replace(/_/g, "/");
    const json =
      typeof atob === "function"
        ? atob(normalized)
        : Buffer.from(normalized, "base64").toString("utf8");
    return JSON.parse(json) as DecodedFluxyJwt;
  } catch {
    return {};
  }
}

export function jwtRefreshDelayMs(expSeconds: number, bufferMs = 5 * 60 * 1000): number {
  return Math.max(expSeconds * 1000 - Date.now() - bufferMs, 0);
}

File: packages/flutter-sdk/lib/src/jwt_utils.dart
import 'dart:convert';

class DecodedFluxyJwt {
  final int? exp;
  final String? sub;
  final String? tid;
  DecodedFluxyJwt({this.exp, this.sub, this.tid});
}

DecodedFluxyJwt decodeFluxyJwtPayload(String token) {
  try {
    final parts = token.split('.');
    if (parts.length < 2) return DecodedFluxyJwt();
    var normalized = parts[1].replaceAll('-', '+').replaceAll('_', '/');
    while (normalized.length % 4 != 0) normalized += '=';
    final json = utf8.decode(base64Decode(normalized));
    final map = jsonDecode(json) as Map<String, dynamic>;
    return DecodedFluxyJwt(
      exp: map['exp'] as int?,
      sub: map['sub'] as String?,
      tid: map['tid'] as String?,
    );
  } catch (_) {
    return DecodedFluxyJwt();
  }
}

int jwtRefreshDelayMs(int expSeconds, {int bufferMs = 300000}) {
  final ms = expSeconds * 1000 - DateTime.now().millisecondsSinceEpoch - bufferMs;
  return ms > 0 ? ms : 0;
}
